habitat-lab/habitat/tasks/nav/nav.py

```diff
@@ -1289,16 +1402,553 @@ def step(
+            next_camera_pitch_ang = self._ang_range_camera_pitch[0]
+            goal_rigid_state.rotation = mn.Quaternion.rotation(
+                mn.Rad(next_camera_pitch_ang), mn.Vector3(1, 0, 0)
+            )
+
+        # Update all the sensors
+        for sensor in sensor_names:
+            self._sim.agents[0]._sensors[  # type: ignore
+                sensor
+            ].node.rotation = goal_rigid_state.rotation
+
+        return self._sim.get_agent_state()
+
+    def _get_agent_observation(self, agent_state=None):
+        position = None
+        rotation = None
+
+        if agent_state is not None:
+            position = agent_state.position
+            rotation = [
+                *agent_state.rotation.vector,
+                agent_state.rotation.scalar,
+            ]
+
+        return self._sim.get_observations_at(
+            position=position,
+            rotation=rotation,
             keep_agent_at_new_pose=True,
         )
 
-        # TODO: Make a better way to flag collisions
-        self._sim._prev_sim_obs["collided"] = collided  # type: ignore
+    def _get_camera_pitch_angle(self):
+        """
+        Get the camera pitch angle
+        """
+        # Get the sensor node
+        sensor_name = list(self._sim.agents[0]._sensors.keys())[0]  # type: ignore
+        sensor_state = self._sim.agents[0]._sensors[sensor_name].node  # type: ignore
+
+        # Construct the sensor rigid state
+        camera_mn_quat = sensor_state.rotation
+
+        # Get the current camera pitch angle
+        camera_rotvec = quaternion.from_float_array(
+            np.array([camera_mn_quat.scalar] + list(camera_mn_quat.vector))
+        )
+        camera_rotvec = quaternion.as_rotation_vector(camera_rotvec)
+        camera_angle = camera_rotvec[0]
+        return camera_angle
+
+    @staticmethod
+    def _scale_inputs(
+        input_val: float, input_range: List[float], output_range: List[float]
+    ) -> float:
+        """
+        Transform input from input range to output range
+        (ex: from normalized input in range [-1, 1] to [y_min, y_max])
+        TODO: This function should go into utils of some sort
+        """
+        w_input = input_range[1] - input_range[0]
+        w_output = output_range[1] - output_range[0]
+        return (
+            output_range[0] + (input_val - input_range[0]) * w_output / w_input
+        )
+
+
+@registry.register_task_action
+class WaypointAction(VelocityAction):
+    name: str = "waypoint_control"
 
-        return agent_observations
+    def __init__(self, *args: Any, **kwargs: Any):
+        super().__init__(*args, **kwargs)
+
+        # Init goto velocity controller
+        try:
+            from habitat.utils.controller_wrapper import (
+                DiffDriveVelocityController,
+            )
+
+            self.w2v_controller = DiffDriveVelocityController(self._config)
+        except ModuleNotFoundError as exc:
+            additional_error_message = """
```

Comment:
```suggestion
            additional_error_message = f"""
            {exc.message}
```

Reply:
We are re-raising the error with `raise ModuleNotFoundError(additional_error_message) from exc`, which should display the original error message. Is there another reason we would need to explicitly print out the error message?